### PR TITLE
Fix docker image reference parsing by DockerImageIdentifierParser

### PR DESCRIPTION
**Pull request by garagatyi**
Parsing used to use reference definition from docker repo, but
fails in some very simple cases.
Rework parsing. Add comments. Add tests.

### CODE REVIEW

**Comment by garagatyi** [comment]
@skabashnyuk @evoevodin @benoitf Please review
@riuvshin Are you OK to merge it in 4.3?

**Comment by skabashnyuk** [comment]
ок

File: plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/parser/DockerImageIdentifierParser.java [hunk 1]
```diff
@@ -26,59 +26,74 @@
...
                 }
             }
         }
 
-        Matcher matcher = IMAGE_PATTERN.matcher(workingCopyOfImage);
-        if (!matcher.matches()) {
-            throw new DockerFileException("Provided image reference is invalid");
+        index = workingCopyOfImage.indexOf('/');
+        String beforeSlash = index > -1 ? workingCopyOfImage.substring(0, index) : "";
+        if (beforeSlash.isEmpty() || (!beforeSlash.contains(".") &&
```
Comment (voievodin) [inline plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/parser/DockerImageIdentifierParser.java]: It is not clear from the context what these `contains(...)` invocations actually check, consider extracting this condition into the method or adding a comment explaining the context of the condition.

File: plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/parser/DockerImageIdentifierParser.java [hunk 2]
```diff
@@ -26,59 +26,74 @@
...
+    private static final String  NAME               = "(?:" + REGISTRY + "/)?" + REPOSITORY;
+    private static final String  REFERENCE          = NAME + "(?::" + TAG + ")?" + "(?:@" + DIGEST + ")?";
+    private static final Pattern IMAGE_PATTERN      = Pattern.compile(REFERENCE);
 
+    /**
+     * Validates and parse docker image reference into object that holds reference components
+     *
+     * @param image image reference to parse
+     * @throws DockerFileException
+     * @throws IllegalArgumentException
```
Comment (voievodin) [inline plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/parser/DockerImageIdentifierParser.java]: There is no `IllegalArgumentException` thrown + `DockerFileExpcetion` is not documented, please do it.

**Comment by voievodin** [comment]
The solution is okay for me, please fix the comments above.

**Comment by riuvshin** [comment]
OK

File: plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/parser/DockerImageIdentifierParser.java (hunk 1 above)
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/parser/DockerImageIdentifierParser.java]: done

File: plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/parser/DockerImageIdentifierParser.java (hunk 2 above)
Comment (garagatyi) [inline plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/parser/DockerImageIdentifierParser.java]: done

**Pull request merged**